### PR TITLE
Propagate schema operations

### PR DESCRIPTION
**Pull request by agedemenli**
DESCRIPTION: Propagates schema operations

This PR adds propagation of schemas as objects. Created new functions like `PreprocessCreateSchemaStmt` and `DeparseCreateSchemaStmts` to achieve this goal. With this PR, now we also propagate `DROP SCHEMA` commands to workers, which will solve Citus' cleanup problem on workers. See #5615

Note that, for transaction blocks, we require sequential mode for `CREATE SCHEMA` statements; otherwise we don't propagate it, as it would cause potential deadlocks. On the other hand, for `DROP` and `ALTER` commands, we just switch to sequential mode, if no parallel query executed in that transaction block. If not, we error out with a message that tells the user to switch to sequential mode.

For non tx block cases, we always propagate schema commands.

fixes: #5615

### CODE REVIEW

**Comment by onderkalaci** [comment]
does dropping schema removes pg_dist_object entries from the worker nodes? Say, there is a type in the schema, and you do `DROP SCHEMA .. CASCADE`. Is the type  removed from worker's pg_dist_object? Would be good to verify that as well

**Comment by agedemenli** [comment]
> does dropping schema removes pg_dist_object entries from the worker nodes? Say, there is a type in the schema, and you do `DROP SCHEMA .. CASCADE`. Is the type removed from worker's pg_dist_object? Would be good to verify that as well

Yes, it gets removed from worker's pg_dist_object as well. I'll add a test for that, once I fix the existing tests

**Comment by agedemenli** [comment]
Self note: Only happens when coordinator is added
Still investigating
```sql
CREATE FUNCTION dummy_function() RETURNS trigger AS $dummy_function$
BEGIN
    NEW.value := value+1;
    RETURN NEW;
END;
$dummy_function$ LANGUAGE plpgsql;
CREATE TABLE dummy_reference_table(a int unique, b int);
INSERT INTO dummy_reference_table SELECT i FROM generate_series(-1, 5) i;
INSERT INTO dummy_reference_table VALUES (99),(100),(599),(600);
SELECT create_reference_table('dummy_reference_table');
create schema "interesting!schema";
CREATE TABLE "interesting!schema"."citus_local!_table"(value int);
ALTER TABLE "interesting!schema"."citus_local!_table" ADD CONSTRAINT fkey_to_dummy_2 FOREIGN KEY (value) REFERENCES dummy_reference_table(a);

CREATE TRIGGER trig
BEFORE INSERT ON "interesting!schema"."citus_local!_table"
FOR EACH STATEMENT EXECUTE FUNCTION dummy_function();

BEGIN;
    CREATE EXTENSION seg;
    -- ALTER TRIGGER DEPENDS ON
    ALTER TRIGGER trig ON "interesting!schema"."citus_local!_table" DEPENDS ON EXTENSION seg;
ERROR:  extension "seg" does not exist
CONTEXT:  while executing command on localhost:9701
```

File: src/backend/distributed/commands/schema.c [hunk 1]
```diff
@@ -53,6 +84,32 @@ PreprocessDropSchemaStmt(Node *node, const char *queryString,
...
+	if (!includesDistributedServer)
+	{
+		return NIL;
+	}
+
+	if (list_length(dropStatement->objects) > 1)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot drop distributed schema with other schemas"),
+						errhint("Try dropping each object in a separate DROP command")));
```
Comment (thanodnl) [inline src/backend/distributed/commands/schema.c:108]: could you explain here, and possibly in a comment on the function why we disallow the dropping of multiple schema's in 1 command?

If you look at the implementation of other drop commands you will see that dropping multiple objects at once is supported.

We achieve this by filtering the list of schema's to drop by the once that are distributed. Before we deparse we use this filtered list, and restore the original list thereafter to locally drop all listed schema's. Any reason that doesn't work here?

File: src/backend/distributed/commands/schema.c [hunk 2]
```diff
@@ -61,11 +118,6 @@ PreprocessDropSchemaStmt(Node *node, const char *queryString,
 	Oid scanIndexId = InvalidOid;
 	bool useIndex = false;
 
-	if (dropStatement->behavior != DROP_CASCADE)
-	{
-		return NIL;
-	}
-
 	Value *schemaValue = NULL;
 	foreach_ptr(schemaValue, dropStatement->objects)
```
Comment (thanodnl) [inline src/backend/distributed/commands/schema.c:122]: I know that this for-loop predates your changes, I have a hard time understanding what we are doing here.

If you have figured out what this loop is for can you do any of the following:
 - leave a comment on top of the loop explaining its purpose
 - extract the loop into a self describing function

If you are not completely sure yourself,  please ignore this comment. It is not required for your change, mostly trying to leave the camping site better then we found it now that we are here.

File: src/backend/distributed/deparser/deparse_schema_stmts.c [hunk 3]
```diff
@@ -53,6 +83,61 @@ DeparseAlterSchemaRenameStmt(Node *node)
...
+	Assert(list_length(stmt->objects) == 1);
+
+	appendStringInfoString(buf, "DROP SCHEMA ");
+
+	if (stmt->missing_ok)
+	{
+		appendStringInfoString(buf, "IF EXISTS ");
+	}
+
+	appendStringInfo(buf, "%s", quote_identifier(strVal(linitial(stmt->objects))));
```
Comment (thanodnl) [inline src/backend/distributed/deparser/deparse_schema_stmts.c:128]: I know you only implemented this for deparsing 1 schema name before, see my comment there.

Even if there is no a good reason to disallow dropping more than 1 schema at a time I'd really love to see this future proved to add all schema's in the statement.

Statement deparsing is reused from many locations which assume that they either work correctly or have Asserts that we only expect 1 schema here.

As implementing correct deparsing is trivial please do so,

```suggestion
     foreach_ptr(...)
     {
              if (!first) appendStringInfoString(buf, ", ");
              appendStringInfoString(buf, quote_identifier(strVal(current_element)));
     }
```

**Review by thanodnl — CHANGES_REQUESTED** [review]

File: src/backend/distributed/commands/schema.c [hunk 4]
```diff
@@ -296,3 +320,101 @@ EnsureSequentialModeForSchemaDDL(void)
...
+
+	if (!IsMultiStatementTransaction())
+	{
+		return true;
+	}
+
+	if (MultiShardConnectionType != SEQUENTIAL_CONNECTION)
+	{
+		return false;
+	}
```
Comment (thanodnl) [inline src/backend/distributed/commands/schema.c:417]: These two statements are highly correlated, I would prefer to have a positive check on the transaction with the sequential check embedded in there to make the logic slightly easier to grasp.

The combination of negating if statements with alternating return values make it _non_ trivial to grasp the logic.

Nevertheless, this could use a bit of comments explaining what we are checking and why.

```suggestion
	if (IsMultiStatementTransaction())
	{
		if (MultiShardConnectionType != SEQUENTIAL_CONNECTION)
		{
			return false;
		}
	}
```

File: src/backend/distributed/commands/schema.c [hunk 5]
```diff
@@ -53,76 +86,44 @@ PreprocessDropSchemaStmt(Node *node, const char *queryString,
...
-				systable_endscan(scanDescriptor);
-				table_close(pgClass, NoLock);
-				return NIL;
-			}
+	EnsureSequentialModeForSchemaDDL();
 
-			heapTuple = systable_getnext(scanDescriptor);
-		}
+	char *schemaName = strVal(linitial(dropStatement->objects));
+	InvalidateFKeysForDistributedTablesInSchema(schemaName);
```
Comment (thanodnl) [inline src/backend/distributed/commands/schema.c:116]: Love this! makes the code way easier to read

File: src/backend/distributed/deparser/deparse_schema_stmts.c [hunk 6]
```diff
@@ -53,6 +84,71 @@ DeparseAlterSchemaRenameStmt(Node *node)
...
+
+	if (stmt->missing_ok)
+	{
+		appendStringInfoString(buf, "IF EXISTS ");
+	}
+
+	Value *schemaValue = NULL;
+	foreach_ptr(schemaValue, stmt->objects)
+	{
+		const char *serverString = quote_identifier(strVal(schemaValue));
```
Comment (thanodnl) [inline src/backend/distributed/deparser/deparse_schema_stmts.c:132]:
```suggestion
		const char *schemaString = quote_identifier(strVal(schemaValue));
```

File: src/include/citus_config.h.in [hunk 7]
```diff
@@ -31,14 +31,17 @@
...
+
 /* Define to 1 if you have the <inttypes.h> header file. */
 #undef HAVE_INTTYPES_H
 
 /* Define to 1 if you have the `curl' library (-lcurl). */
 #undef HAVE_LIBCURL
 
 /* Define to 1 if you have the `lz4' library (-llz4). */
-#undef HAVE_CITUS_LIBLZ4
+#undef HAVE_LIBLZ4
```
Comment (thanodnl) [inline src/include/citus_config.h.in:44]: These changes seem unrelated to your PR, lets not change the configure related files when we are not changing anything to the automake files.

File: src/backend/distributed/commands/schema.c (hunk 1 above)
Comment (thanodnl) [reply src/backend/distributed/commands/schema.c:110]: This is a big outstanding question before we can merge the PR.

File: src/test/regress/expected/columnar_truncate.out [hunk 8]
```diff
@@ -268,6 +268,13 @@ set columnar.compression to default;
 CREATE USER truncate_user;
 NOTICE:  not propagating CREATE ROLE/USER commands to worker nodes
 HINT:  Connect to worker nodes directly to manually create all necessary users and roles.
+SELECT run_command_on_workers($$CREATE USER truncate_user;$$);
+      run_command_on_workers
+---------------------------------------------------------------------
+ (localhost,57637,t,"CREATE ROLE")
+ (localhost,57638,t,"CREATE ROLE")
+(2 rows)
+
```
Comment (thanodnl) [inline src/test/regress/expected/columnar_truncate.out:277]: I was checking this as one of the few enterprise merge conflict.

In general it will be easier to create the remote users before the local users with regards to enterprise changes required for this change.

When enterprise propagates the user it is expected to be idempotent, meaning if the user already exists it configures it like it is supposed to be without error. You calling this after the remote user might be created in enterprise will return an error from the workers which changes output files here.

I will suggest to move this statement above the local creation.

File: src/backend/distributed/commands/schema.c [hunk 9]
```diff
@@ -53,76 +85,48 @@ PreprocessDropSchemaStmt(Node *node, const char *queryString,
...
-		{
-			continue;
-		}
+	EnsureSequentialModeForSchemaDDL();
 
-		pgClass = table_open(RelationRelationId, AccessShareLock);
+	Value *schemaVal = NULL;
+	foreach_ptr(schemaVal, distributedSchemas)
+	{
+		InvalidateFKeysForDistributedTablesInSchema(strVal(schemaVal));
```
Comment (thanodnl) [inline src/backend/distributed/commands/schema.c:109]: Looking at the code this replaces on the original line 115 we had an early escape once the graph has been invalidated. With multiple schema's do we need to invalidate the graph twice?

Has this been considered as an option?

I don't think it is a huge performance problem, given this is non-frequent executed code. If the readability decreases we can keep it like this, but a comment that this behaviour is expected would be nice.

Alternatively this could be refactored slightly more in a function that simply checks if the schema has any tables referencing something here. When it does an early escape with a bool will suffice.

Once we have any schema that has references that require invalidation we can simply call `MarkInvalidateForeignKeyGraph` from here.

This all is under the assumption that the original code was _only_ having the side effect of invalidating the fkey graph.

As said earlier, this is by no means required, but reasoning why we don't do this will help our future selfs.

File: src/backend/distributed/commands/schema.c [hunk 10]
```diff
@@ -296,3 +323,78 @@ EnsureSequentialModeForSchemaDDL(void)
...
+		if (relationId == InvalidOid || !IsCitusTable(relationId))
+		{
+			heapTuple = systable_getnext(scanDescriptor);
+			continue;
+		}
+
+		/* invalidate foreign key cache if the table involved in any foreign key */
+		if (TableReferenced(relationId) || TableReferencing(relationId))
+		{
+			MarkInvalidateForeignKeyGraph();
```
Comment (thanodnl) [inline src/backend/distributed/commands/schema.c:372]: Same as earlier, the original code had an early escape after the graph had been invalidated. Any reason we have changed that behaviour?

File: src/backend/distributed/commands/schema.c [hunk 11]
```diff
@@ -53,76 +85,48 @@ PreprocessDropSchemaStmt(Node *node, const char *queryString,
...
-				systable_endscan(scanDescriptor);
-				table_close(pgClass, NoLock);
-				return NIL;
-			}
+	/* modify object list such that it only contains the distribute ones */
+	dropStatement->objects = distributedSchemas;
 
-			heapTuple = systable_getnext(scanDescriptor);
-		}
+	/* deparse sql with the distributed ones */
```
Comment (thanodnl) [inline src/backend/distributed/commands/schema.c:118]: this comment is exactly telling what the code tells as well.

In general you don't have to document what it does, only why it does so if that is not obvious.
I would suggest removing this comment as it doesn't add value

**Review by thanodnl — COMMENTED** [review]
I love how many lines we could remove with this PR! Most of them in output files of test suites ofc, but still a net win!

high level comments
 - I see a lot of DROP SCHEMA's that used to be in 1 statement being split out. Guess this is due to the late addition of support for multiple drops. Would be great to undo those changes. As I left on one of the occurrences, this is apparently more common than I anticipated. Good to have it supported now
 - Bit of change in behaviour in the fkgey graph invalidation. Either comment on the PR why, or refactor as suggested if that makes the code more readable.
 - Some general hints on how to prevent mayham when merging to enterprise

File: src/backend/distributed/commands/schema.c [hunk 12]
```diff
@@ -296,3 +329,90 @@ EnsureSequentialModeForSchemaDDL(void)
...
+ */
+static bool
+ShouldPropagateCreateSchemaStmt()
+{
+	if (!ShouldPropagate())
+	{
+		return false;
+	}
+
+	if (IsMultiStatementTransaction())
```
Comment (onderkalaci) [inline src/backend/distributed/commands/schema.c:409]: Given that we plan to use this block in other places as well, I'd suggest:

```C
if (IsMultiStatementTransaction() && MultiShardConnectionType != SEQUENTIAL_CONNECTION)
{
   return false;
}
```

File: src/backend/distributed/commands/schema.c [hunk 13]
```diff
@@ -53,76 +85,54 @@ PreprocessDropSchemaStmt(Node *node, const char *queryString,
...
 
-		if (namespaceOid == InvalidOid)
+	Value *schemaVal = NULL;
+	foreach_ptr(schemaVal, distributedSchemas)
+	{
+		if (SchemaHasDistributedTableWithFKey(strVal(schemaVal)))
 		{
-			continue;
+			MarkInvalidateForeignKeyGraph();
+			break;
```
Comment (thanodnl) [inline src/backend/distributed/commands/schema.c:112]: ❤️

**Review by thanodnl — APPROVED** [review]
